### PR TITLE
improvement: Use LRU instead of maps on defaultStore

### PR DESCRIPTION
**Pull request by ajnavarro**
# Description

Changing actual cache implementation to use an LRU cache improves resource usage and avoids exhaustion.

### CODE REVIEW

**Comment by tbruyelle** [comment]
Can you explain why maps are exhausting plz ?

**Comment by moul** [comment]
Can you provide some metrics, i.e., benchmark before/after?

cc @peter7891 good example of candidate for the performance management framework.

File: go.mod [hunk 1]
```diff
@@ -55,6 +55,7 @@ require (
 	github.com/google/flatbuffers v1.12.1 // indirect
 	github.com/gorilla/securecookie v1.1.1 // indirect
 	github.com/gorilla/sessions v1.2.1 // indirect
+	github.com/hashicorp/golang-lru/v2 v2.0.2 // indirect
```
Comment (tbruyelle) [inline go.mod:58]: Shouldn't be `indirect`, please run `go mod tidy` to fix that.

File: gnovm/pkg/gnolang/store.go [hunk 2]
```diff
@@ -6,12 +6,17 @@ import (
...
+
 	"github.com/gnolang/gno/tm2/pkg/amino"
 	"github.com/gnolang/gno/tm2/pkg/std"
 	"github.com/gnolang/gno/tm2/pkg/store"
 )
 
 const iavlCacheSize = 1024 * 1024 // TODO increase and parameterize.
+const objectCacheSize = 100       // TODO parameterize.
+const typeCacheSize = 1000        // TODO parameterize.
+const nodeCacheSize = 10000       // TODO parameterize.
```
Comment (tbruyelle) [inline gnovm/pkg/gnolang/store.go:19]: How did you evaluate those values ?

File: go.mod (hunk 1 above)
Comment (ajnavarro) [reply go.mod]: done

File: gnovm/pkg/gnolang/store.go (hunk 2 above)
Comment (ajnavarro) [reply gnovm/pkg/gnolang/store.go:19]: Executing a stress test using supernova. Ideally, we should define a total max memory available for store cache.

**Comment by ajnavarro** [comment]
> @ajnavarro, thank you, can you provide some metrics, i.e., benchmark before/after?
> 
> Warn: new dependency, to be reviewed.
> 
> cc @peter7891 good example of candidate for the performance management framework.

I can create some benchmarks measuring the total memory used.

**Comment by ajnavarro** [comment]
@moul @tbruyelle I think I found a bug. After limiting the number of objects allowed to be on memory, some tests started to fail saying that a package is missing from the store (it should be there if it was on the cache before). That might be considered a bug. WDYT?

**Comment by tbruyelle** [comment]
> @moul @tbruyelle I think I found a bug. After limiting the number of objects allowed to be on memory, some tests started to fail saying that a package is missing from the store (it should be there if it was on the cache before). That might be considered a bug. WDYT?

Sure, what is the error exactly ?

**Comment by tbruyelle** [comment]
OK so to be sure I understand well, some packages (I assume the ones added via `SetCachePackage()`) aren't retrievable from the store via `loadObjectSafe()` right ?

**Comment by ajnavarro** [comment]
Hmm, `loadObjectSafe()` was never able to obtain packages stored using the `SetCachePackage()` method.

The problem with the actual change is that `SetCachePackage` is using an in-memory cache that was not expecting to be evicted in any case, but with that change adding an LRU cache we are discarding less used packages from memory.

`GetPackage` is looking for packages from memory and from other two sources if they are present: baseStore and pkgGetter. With this change, we are removing some packages from memory, so it is not able to find them.

**Comment by ajnavarro** [comment]
I asked @zivkovicmilos to make this PR a draft because I'll need more time to think about throwaway packages.

**Pull request closed without merging**